src/model_theory/basic.lean

```diff
@@ -545,6 +558,247 @@ lemma closure_union (s t : set M) : closure L (s ∪ t) = closure L s ⊔ closur
 lemma closure_Union {ι} (s : ι → set M) : closure L (⋃ i, s i) = ⨆ i, closure L (s i) :=
 (substructure.gi L M).gc.l_supr
 
+/-!
+### `comap` and `map`
+-/
+
+/-- The preimage of a substructure along a homomorphism is a substructure. -/
+def comap (φ : M →[L] N) (S : L.substructure N) : L.substructure M :=
+{ carrier := (φ ⁻¹' S),
+  fun_mem := λ n f x hx, begin
+    rw [mem_preimage, φ.map_fun],
+    exact S.fun_mem f (φ ∘ x) hx,
+  end }
+
+@[simp]
+lemma coe_comap (S : L.substructure N) (f : M →[L] N) : (S.comap f : set M) = f ⁻¹' S := rfl
+
+@[simp]
+lemma mem_comap {S : L.substructure N} {f : M →[L] N} {x : M} : x ∈ S.comap f ↔ f x ∈ S := iff.rfl
+
+
+lemma comap_comap (S : L.substructure P) (g : N →[L] P) (f : M →[L] N) :
+  (S.comap g).comap f = S.comap (g.comp f) :=
+rfl
+
+@[simp]
+lemma comap_id (S : L.substructure P) : S.comap (hom.id _ _) = S :=
+ext (by simp)
+
+/-- The image of a substructure along a homomorphism is a substructure. -/
+def map (φ : M →[L] N) (S : L.substructure M) : L.substructure N :=
+{ carrier := (φ '' S),
+  fun_mem := λ n f x hx, (mem_image _ _ _).1 ⟨fun_map f (λ i, classical.some (hx i)),
+      S.fun_mem f _ (λ i, (classical.some_spec (hx i)).1), begin
+    simp only [hom.map_fun, set_like.mem_coe],
+    exact congr rfl (funext (λ i, (classical.some_spec (hx i)).2)),
+  end⟩ }
+
+@[simp]
+lemma coe_map (f : M →[L] N) (S : L.substructure M) :
+  (S.map f : set N) = f '' S := rfl
+
+@[simp]
+lemma mem_map {f : M →[L] N} {S : L.substructure M} {y : N} :
+  y ∈ S.map f ↔ ∃ x ∈ S, f x = y :=
+mem_image_iff_bex
+
+
+lemma mem_map_of_mem (f : M →[L] N) {S : L.substructure M} {x : M} (hx : x ∈ S) : f x ∈ S.map f :=
+mem_image_of_mem f hx
+
+
+lemma apply_coe_mem_map (f : M →[L] N) (S : L.substructure M) (x : S) : f x ∈ S.map f :=
+mem_map_of_mem f x.prop
+
+
+lemma map_map (g : N →[L] P) (f : M →[L] N) : (S.map f).map g = S.map (g.comp f) :=
+set_like.coe_injective $ image_image _ _ _
+
+
```

Comment:
There seem to be a bunch of double empty lines between the lemma declarations here. Is that intentional?

Reply:
Could you give a motivation?